packages/dev/core/src/Meshes/abstractMesh.ts

```diff
@@ -43,6 +43,118 @@ import type { Collider } from "../Collisions/collider";
 import type { TrianglePickingPredicate } from "../Culling/ray";
 import type { RenderingGroup } from "../Rendering/renderingGroup";
 import type { IEdgesRendererOptions } from "../Rendering/edgesRenderer";
+import type { MorphTarget } from "../Morph/morphTarget";
+import { nativeOverride } from "../Misc/decorators";
+
+/** @internal */
+function applyMorph(data: FloatArray, kind: string, morphTargetManager: MorphTargetManager): void {
+    let getTargetData: Nullable<(target: MorphTarget) => Nullable<FloatArray>> = null;
+    switch (kind) {
+        case VertexBuffer.PositionKind:
+            getTargetData = (target) => target.getPositions();
+            break;
+        case VertexBuffer.NormalKind:
+            getTargetData = (target) => target.getNormals();
+            break;
+        case VertexBuffer.TangentKind:
+            getTargetData = (target) => target.getTangents();
+            break;
+        case VertexBuffer.UVKind:
+            getTargetData = (target) => target.getUVs();
+            break;
+        default:
+            return;
+    }
+
+    for (let index = 0; index < data.length; index++) {
+        let value = data[index];
+        for (let targetCount = 0; targetCount < morphTargetManager.numTargets; targetCount++) {
+            const target = morphTargetManager.getTarget(targetCount);
+            const influence = target.influence;
+            if (influence !== 0) {
+                const targetData = getTargetData(target);
+                if (targetData) {
+                    value += (targetData[index] - data[index]) * influence;
+                }
+            }
+        }
+        data[index] = value;
+    }
+}
+
+/** @internal */
+function applySkeleton(
+    data: FloatArray,
+    kind: string,
+    skeletonMatrices: Float32Array,
+    matricesIndicesData: FloatArray,
+    matricesWeightsData: FloatArray,
+    matricesIndicesExtraData: Nullable<FloatArray>,
+    matricesWeightsExtraData: Nullable<FloatArray>
+): void {
+    const tempVector = TmpVectors.Vector3[0];
+    const finalMatrix = TmpVectors.Matrix[0];
+    const tempMatrix = TmpVectors.Matrix[1];
+
+    const transformFromFloatsToRef = kind === VertexBuffer.NormalKind ? Vector3.TransformNormalFromFloatsToRef : Vector3.TransformCoordinatesFromFloatsToRef;
+
+    for (let index = 0, matWeightIdx = 0; index < data.length; index += 3, matWeightIdx += 4) {
+        finalMatrix.reset();
+
+        let inf: number;
+        let weight: number;
+        for (inf = 0; inf < 4; inf++) {
+            weight = matricesWeightsData[matWeightIdx + inf];
+            if (weight > 0) {
+                Matrix.FromFloat32ArrayToRefScaled(skeletonMatrices, Math.floor(matricesIndicesData[matWeightIdx + inf] * 16), weight, tempMatrix);

```

Comment:
I'm not sure why the existing code is doing a `floor` here...

Reply:
I questioned this also, but I'm guessing it can prevent bad data?